mbed_project/mbed_program.py

```diff
@@ -120,42 +114,43 @@ def from_existing_local_program_directory(cls, dir_path: Path) -> "MbedProgram":
-        repo = git.Repo(str(program_root))
+        repo = git_utils.git.Repo(str(program_root))
         program = MbedProgramData.from_existing(program_root)
         mbed_os = MbedOS.from_existing(program_root / MBED_OS_DIR_NAME)
         return cls(repo, program, mbed_os)
 
     def resolve_libraries(self) -> None:
-        """Clone all external dependencies defined in .lib files.
+        """Resolve all external dependencies defined in .lib files."""
+        self.lib_references.resolve()
 
-        This function will recursively clone all unresolved dependencies until all are resolved.
+    def checkout_libraries(self, force: bool = False) -> None:
+        """Check out all resolved libraries to revisions specified in .lib files."""
+        self.lib_references.checkout(force)
 
-        Raises:
-            VersionControlError: The git clone failed.
-        """
-        for lib in self._find_unresolved_lib_references():
-            git_ref = lib.get_git_reference()
-            logger.info(f"Resolving library reference {git_ref.repo_url}.")
-            _clone_lib(git_ref, lib.source_code_path)
+    def list_known_library_dependencies(self) -> List[str]:
+        """Returns a list of all known library dependencies."""
+        return [lib.get_git_reference().repo_url for lib in self.lib_references.iter_all()]
 
-        # Check if we find any new references after cloning dependencies.
-        if list(self._find_unresolved_lib_references()):
-            self.resolve_libraries()
+    def has_unresolved_libraries(self) -> bool:
+        """Checks if any unresolved library dependencies exist in the program tree."""
+        return bool(list(self.lib_references.iter_unresolved()))
 
-    def _find_lib_references(self) -> Iterable[MbedLibReference]:
-        program_root = self.metadata.mbed_file.parent
-        return (
-            MbedLibReference(lib, lib.with_suffix(""))
-            for lib in program_root.rglob("*.lib")
-            if not self._in_ignore_path(lib)
-        )
 
-    def _find_unresolved_lib_references(self) -> Iterable[MbedLibReference]:
-        return (lib for lib in self._find_lib_references() if not lib.is_resolved())
+def parse_url(name_or_url: str) -> Dict[str, str]:
+    """Create a valid github/armmbed url from a program name.
 
-    def _in_ignore_path(self, lib_reference_path: Path) -> bool:
-        """Mbed OS contains some test .lib files we want to ignore."""
-        return self.mbed_os.root in lib_reference_path.parents
+    Args:
+        url: The URL, or a program name to turn into an URL.
+
+    Returns:
+        Dictionary containing the remote url and the destination path for the clone.
+    """
+    url_obj = urlparse(name_or_url)
+    if url_obj.hostname:
+        url = url_obj.geturl()
+    else:
+        url = f"https://github.com/armmbed/{url_obj.path}"
+    return {"url": url, "dst_path": url_obj.path.rsplit("/", maxsplit=1)[-1].replace("/", "")}
```

Comment:
May be worth just adding a comment explaining why the string processing is for.

Reply:
Yeah I'll add a comment here.

Reply:
Can this handle spaces and does it need to?

Reply:
> Can this handle spaces and does it need to?

I don't think it needs to, the github URLS can't have spaces in them AFAIK.